Fetching `QueryResults` from a query having both `groupBy` and `having` clause results in a `QuerySyntaxException`

Fetching `QueryResults` from a query having both `groupBy` and `having` clause results in a `QuerySyntaxException`. Here are two test cases, the first one executes without error the second results in the above described error: 

```
    @Test
    public void groupByHavingFetch() {
        QCustomer customer = QCustomer.customer;
        JPAQuery<Customer> customerQuery = new JPAQuery<Customer>(entityManager, HQLTemplates.DEFAULT).from(customer);

        customerQuery.groupBy(customer.lastName).having(customer.id.count().gt(1)).fetch();
    }

    @Test
    public void groupByHavingFetchResults() {
        QCustomer customer = QCustomer.customer;
        JPAQuery<Customer> customerQuery = new JPAQuery<Customer>(entityManager, HQLTemplates.DEFAULT).from(customer);

        customerQuery.groupBy(customer.lastName).having(customer.id.count().gt(1)).fetchResults();
    }
```
The test in error generates following error:

```
java.lang.IllegalArgumentException: org.hibernate.hql.internal.ast.QuerySyntaxException: unexpected token: having near line 3, column 1 [select count(distinct customer.lastName)
from Customer customer
having count(customer.id) > ?1]

	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:131)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:155)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:162)
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:662)
	at org.hibernate.internal.AbstractSessionImpl.createQuery(AbstractSessionImpl.java:23)
	at com.querydsl.jpa.impl.AbstractJPAQuery.createQuery(AbstractJPAQuery.java:101)
	at com.querydsl.jpa.impl.AbstractJPAQuery.fetchResults(AbstractJPAQuery.java:211)
	at GroupByHavingTest.groupByHavingFetchResults(GroupByHavingTest.java:49)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:237)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
Caused by: org.hibernate.hql.internal.ast.QuerySyntaxException: unexpected token: having near line 3, column 1 [select count(distinct customer.lastName)
from Customer customer
having count(customer.id) > ?1]
	at org.hibernate.hql.internal.ast.QuerySyntaxException.convert(QuerySyntaxException.java:74)
	at org.hibernate.hql.internal.ast.ErrorCounter.throwQueryException(ErrorCounter.java:91)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.parse(QueryTranslatorImpl.java:288)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.doCompile(QueryTranslatorImpl.java:187)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.compile(QueryTranslatorImpl.java:142)
	at org.hibernate.engine.query.spi.HQLQueryPlan.<init>(HQLQueryPlan.java:115)
	at org.hibernate.engine.query.spi.HQLQueryPlan.<init>(HQLQueryPlan.java:77)
	at org.hibernate.engine.query.spi.QueryPlanCache.getHQLQueryPlan(QueryPlanCache.java:153)
	at org.hibernate.internal.AbstractSharedSessionContract.getQueryPlan(AbstractSharedSessionContract.java:545)
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:654)
	... 32 more
``` 
The resulting hql Statement is missing the `groupBy` clause.

--- a/querydsl-core/src/main/java/com/querydsl/core/Fetchable.java
+++ b/querydsl-core/src/main/java/com/querydsl/core/Fetchable.java
@@ -78,13 +78,20 @@ public interface Fetchable<T> {
      * {@link QueryResults#getLimit()}, because it will be more performant. Also, count queries cannot be
      * properly generated for all dialects. For example: in JPA count queries can't be generated for queries
      * that have multiple group by expressions or a having clause.
+     * Get the projection in {@link QueryResults} form.
+     *
+     * Use {@link #fetch()} instead if you do not need the total count of rows in the query result.
      *
      * @return results
+     * @see #fetch()
      */
     QueryResults<T> fetchResults();
 
     /**
      * Get the count of matched elements
+     *
+     * Note: not all QueryDSL modules might optimize fetchCount using a count query.
+     * An implementation is allowed to fall back to {@code fetch().size()}.
      *
      * @return row count
      */

--- a/querydsl-jpa/src/main/java/com/querydsl/jpa/impl/AbstractJPAQuery.java
+++ b/querydsl-jpa/src/main/java/com/querydsl/jpa/impl/AbstractJPAQuery.java
@@ -75,9 +75,42 @@ public abstract class AbstractJPAQuery<T, Q extends AbstractJPAQuery<T, Q>> exte
         this.entityManager = em;
     }
 
+    /**
+     * {@inheritDoc}
+     *
+     * @deprecated {@code fetchCount} requires a count query to be computed. In {@code querydsl-sql}, this is done
+     * by wrapping the query in a subquery, like so: {@code SELECT COUNT(*) FROM (&lt;original query&gt;)}. Unfortunately,
+     * JPQL - the query language of JPA - does not allow queries to project from subqueries. As a result there isn't a
+     * universal way to express count queries in JPQL. Historically QueryDSL attempts at producing a modified query
+     * to compute the number of results instead.
+     *
+     * However, this approach only works for simple queries. Specifically
+     * queries with multiple group by clauses and queries with a having clause turn out to be problematic. This is because
+     * {@code COUNT(DISTINCT a, b, c)}, while valid SQL in most dialects, is not valid JPQL. Furthermore, a having
+     * clause may refer select elements or aggregate functions and therefore cannot be emulated by moving the predicate
+     * to the where clause instead.
+     *
+     * In order to support {@code fetchCount} for queries with multiple group by elements or a having clause, we
+     * generate the count in memory instead. This means that the method simply falls back to returning the size of
+     * {@link #fetch()}. For large result sets this may come at a severe performance penalty.
+     *
+     * For very specific domain models where {@link #fetchCount()} has to be used in conjunction with complex queries
+     * containing multiple group by elements and/or a having clause, we recommend using the
+     * <a href="https://persistence.blazebit.com/documentation/1.5/core/manual/en_US/index.html#querydsl-integration">Blaze-Persistence</a>
+     * integration for QueryDSL. Among other advanced query features, Blaze-Persistence makes it possible to select
+     * from subqueries in JPQL. As a result the {@code BlazeJPAQuery} provided with the integration, implements
+     * {@code fetchCount} properly and always executes a proper count query.
+     */
     @Override
+    @Deprecated
     public long fetchCount() {
         try {
+            if (getMetadata().getGroupBy().size() > 1 || getMetadata().getHaving() != null) {
+                logger.warn("Fetchable#fetchCount() was computed in memory! See the Javadoc for AbstractJPAQuery#fetchCount for more details.");
+                Query query = createQuery(null, false);
+                return query.getResultList().size();
+            }
+
             Query query = createQuery(null, true);
             return (Long) query.getSingleResult();
         } finally {
@@ -215,13 +248,53 @@ public abstract class AbstractJPAQuery<T, Q extends AbstractJPAQuery<T, Q>> exte
         }
     }
 
+    /**
+     * {@inheritDoc}
+     *
+     * @deprecated {@code fetchResults} requires a count query to be computed. In {@code querydsl-sql}, this is done
+     * by wrapping the query in a subquery, like so: {@code SELECT COUNT(*) FROM (&lt;original query&gt;)}. Unfortunately,
+     * JPQL - the query language of JPA - does not allow queries to project from subqueries. As a result there isn't a
+     * universal way to express count queries in JPQL. Historically QueryDSL attempts at producing a modified query
+     * to compute the number of results instead.
+     *
+     * However, this approach only works for simple queries. Specifically
+     * queries with multiple group by clauses and queries with a having clause turn out to be problematic. This is because
+     * {@code COUNT(DISTINCT a, b, c)}, while valid SQL in most dialects, is not valid JPQL. Furthermore, a having
+     * clause may refer select elements or aggregate functions and therefore cannot be emulated by moving the predicate
+     * to the where clause instead.
+     *
+     * In order to support {@code fetchResults} for queries with multiple group by elements or a having clause, we
+     * generate the count in memory instead. This means that the method simply falls back to returning the size of
+     * {@link #fetch()}. For large result sets this may come at a severe performance penalty.
+     *
+     * For very specific domain models where {@link #fetchResults()} has to be used in conjunction with complex queries
+     * containing multiple group by elements and/or a having clause, we recommend using the
+     * <a href="https://persistence.blazebit.com/documentation/1.5/core/manual/en_US/index.html#querydsl-integration">Blaze-Persistence</a>
+     * integration for QueryDSL. Among other advanced query features, Blaze-Persistence makes it possible to select
+     * from subqueries in JPQL. As a result the {@code BlazeJPAQuery} provided with the integration, implements
+     * {@code fetchResults} properly and always executes a proper count query.
+     *
+     * Mind that for any scenario where the count is not strictly needed separately, we recommend to use {@link #fetch()}
+     * instead.
+     */
     @Override
+    @Deprecated
     public QueryResults<T> fetchResults() {
         try {
+            QueryModifiers modifiers = getMetadata().getModifiers();
+            if (getMetadata().getGroupBy().size() > 1 || getMetadata().getHaving() != null) {
+                logger.warn("Fetchable#fetchResults() was computed in memory! See the Javadoc for AbstractJPAQuery#fetchResults for more details.");
+                Query query = createQuery(null, false);
+                @SuppressWarnings("unchecked")
+                List<T> resultList = query.getResultList();
+                int offset = modifiers.getOffsetAsInteger() == null ? 0 : modifiers.getOffsetAsInteger();
+                int limit = modifiers.getLimitAsInteger() == null ? resultList.size() : modifiers.getLimitAsInteger();
+                return new QueryResults<T>(resultList.subList(offset, Math.min(resultList.size(), offset + limit)), modifiers, resultList.size());
+            }
+
             Query countQuery = createQuery(null, true);
             long total = (Long) countQuery.getSingleResult();
             if (total > 0) {
-                QueryModifiers modifiers = getMetadata().getModifiers();
                 Query query = createQuery(modifiers, false);
                 @SuppressWarnings("unchecked")
                 List<T> list = (List<T>) getResultList(query);

--- a/querydsl-jpa/src/test/java/com/querydsl/jpa/JPABase.java
+++ b/querydsl-jpa/src/test/java/com/querydsl/jpa/JPABase.java
@@ -29,6 +29,7 @@ import javax.persistence.EntityManager;
 import javax.persistence.FlushModeType;
 import javax.persistence.LockModeType;
 
+import com.querydsl.core.QueryResults;
 import com.querydsl.core.Tuple;
 import org.junit.ClassRule;
 import org.junit.Ignore;
@@ -278,5 +279,26 @@ public class JPABase extends AbstractJPATest implements JPATest {
         for (String row : rows) {
             assertNotNull(row);
         }
+    }
+
+    @Test
+    public void fetchCountResultsGroupByWithMultipleFields() {
+        QueryResults<Tuple> results = query().from(cat)
+                .groupBy(cat.alive, cat.breed)
+                .select(cat.alive, cat.breed, cat.id.sum())
+                .fetchResults();
+
+        assertEquals(1, results.getTotal());
+    }
+
+    @Test
+    public void fetchCountResultsGroupByWithHaving() {
+        QueryResults<Tuple> results = query().from(cat)
+                .groupBy(cat.alive)
+                .having(cat.id.sum().gt(5))
+                .select(cat.alive, cat.id.sum())
+                .fetchResults();
+
+        assertEquals(1, results.getTotal());
     }
 }

